command/agent/retry_join.go

```diff
@@ -114,17 +169,14 @@ func (r *retryJoiner) RetryJoin(serverJoin *ServerJoin) {
 		var err error
 
 		for _, addr := range serverJoin.RetryJoin {
-			switch {
-			case strings.HasPrefix(addr, "provider="):
-				servers, err := r.discover.Addrs(addr, standardLogger)
-				if err != nil {
-					r.logger.Error("determining join addresses failed", "error", err)
-				} else {
-					addrs = append(addrs, servers...)
-				}
-			default:
+			servers, err := r.discover.Addrs(addr, standardLogger)
+			if err != nil {
+				r.logger.Warn("discovering join addresses failed", "join_config", addr, "error", err)
```

Comment:
This is also expected behavior and probably should not be a warn level log. This is Debug-worth at best, and is perhaps not necessary. It is here for debugging during development.

Reply:
Additionally, nothing has necessarily "failed" at this point if `cfg` is an IP addresses, which is valid, then this error block simply means that "discovery" has yielded no results, not that `cfg` is some how invalid.

Reply:
Given we're pushing some of this block's existing behavior down into the `Addrs` method, this feels like an opportunity to be more thoughtful about what we return as an error here. I would expect to see an error if:
* The discovery config is invalid.
* The discovery config is valid but behaved incorrectly. For example, when the `go-netaddrs` `exec=` fails or the `provider=` credentials are bad.

Reply:
>  or the `provider=` credentials are bad.

Could you clarify what "bad" credentials are and how one might verify badness?

Reply:
Sorry, I should have been a little more clear that I'm saying we should treat those two different categories of error differently. 

* The first category is where Nomad can know the configuration is invalid. Ex. the config is `"exec="` but without arguments. In that case, we can return a specific error to the user telling them to go fix their config.
* But the second case is where the failure is outside of Nomad's control. For example, the command we execute fails or you're using `provider=aws` and the IAM policy doesn't actually work. We can't "verify badness" there, so we'd just bubble up the errors we got.

Reply:
Gotcha thanks. I believe the latest commit to this PR does exactly what you describe, but let me know if I've misunderstood your intention here. 

Category 1: https://github.com/hashicorp/nomad/pull/18745/files#diff-3cedac9fc5d67acba9b51036e3dce4d1d42d9e468c213124b0565069324f25a9R95
Category 2: https://github.com/hashicorp/nomad/pull/18745/files#diff-3cedac9fc5d67acba9b51036e3dce4d1d42d9e468c213124b0565069324f25a9R119

Reply:
Yeah this is great :+1: